File: tests/oracle/op_insert_1204.cc
#include "bfd_arch.hh"
#include "libdecomp.hh"

#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

struct Fixture {
  Funcdata *fd;
  BlockBasic *block;
  vector<PcodeOp *> ops;
  map<PcodeOp *,string> names;
  PcodeOp *tracked;
  Varnode *trackedInput;
  Varnode *trackedOutput;

  explicit Fixture(Funcdata *func)
    : fd(func), block((BlockBasic *)0), tracked((PcodeOp *)0),
      trackedInput((Varnode *)0), trackedOutput((Varnode *)0)
  {
    BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
    block = graph.newBlockBasic(fd);
  }

  PcodeOp *make(const string &name,OpCode opcode,int4 inputs=0)
  {
    AddrSpace *codeSpace = fd->getArch()->getDefaultCodeSpace();
    PcodeOp *op = fd->newOp(inputs,Address(codeSpace,0x1000));
    fd->opSetOpcode(op,opcode);
    ops.push_back(op);
    names[op] = name;
    return op;
  }

  string name(PcodeOp *op) const
  {
    map<PcodeOp *,string>::const_iterator iter = names.find(op);
    if (iter == names.end()) throw std::runtime_error("unknown fixture op");
    return (*iter).second;
  }

  string blockOrder(void) const
  {
    ostringstream out;
    bool first = true;
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();iter!=block->endOp();++iter) {
      if (!first) out << ',';
      first = false;
      out << name(*iter);
    }
    return out.str();
  }

  string bankOrder(bool alive) const
  {
    ostringstream out;
    bool first = true;
    list<PcodeOp *>::const_iterator iter = alive ? fd->beginOpAlive() : fd->beginOpDead();
    list<PcodeOp *>::const_iterator enditer = alive ? fd->endOpAlive() : fd->endOpDead();
    for(;iter!=enditer;++iter) {
      if (names.find(*iter) == names.end()) continue;
      if (!first) out << ',';
      first = false;
      out << name(*iter);
    }
    return out.str();
  }

  string parentOrder(void) const
  {
    ostringstream out;
    bool first = true;
    for(vector<PcodeOp *>::const_iterator iter=ops.begin();iter!=ops.end();++iter) {
      if (!first) out << ',';
      first = false;
      out << name(*iter) << ':';
      if ((*iter)->getParent() == block)
        out << block->getIndex();
      else
        out << -1;
    }
    return out.str();
  }

  string opState(void) const
  {
    ostringstream out;
    bool first = true;
    for(vector<PcodeOp *>::const_iterator iter=ops.begin();iter!=ops.end();++iter) {
      PcodeOp *op = *iter;
      if (!first) out << ',';
      first = false;
      out << name(op) << ':' << op->getEvalType()
          << '/' << op->isBranch()
          << '/' << op->isCall()
          << '/' << op->isMarker()
          << '/' << op->isBoolOutput()
          << '/' << op->isFlowBreak()
          << '/' << op->isCodeRef()
          << '/' << op->notPrinted();
    }
    return out.str();
  }

  void dump(const string &stage,bool includeOrder) const
  {
    int4 parent = tracked->getParent() == block ? block->getIndex() : -1;
    int4 descendCount = std::distance(trackedInput->beginDescend(),trackedInput->endDescend());
    std::cout << stage
              << "|block=" << blockOrder()
              << "|alive=" << bankOrder(true)
              << "|dead=" << bankOrder(false)
              << "|parents=" << parentOrder()
              << "|opstate=" << opState()
              << "|flags=" << block->getFlags()
              << "|tracked_dead=" << tracked->isDead()
              << "|tracked_parent=" << parent
              << "|tracked_input_desc=" << descendCount
              << "|tracked_output_def=" << (trackedOutput->getDef() == tracked);
    if (includeOrder)
      std::cout << "|tracked_order=" << tracked->getSeqNum().getOrder();
    std::cout << '\n';
  }
};

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary,"default",&std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found in the BFD symbol table");

    Fixture fixture(fd);
    PcodeOp *terminal = fixture.make("terminal",CPUI_BRANCHIND);
    PcodeOp *target = fixture.make("target",CPUI_STORE);
    PcodeOp *head = fixture.make("head",CPUI_COPY);
    PcodeOp *phiA = fixture.make("phi_a",CPUI_MULTIEQUAL);
    PcodeOp *phiB = fixture.make("phi_b",CPUI_MULTIEQUAL);
    PcodeOp *afterPhi = fixture.make("after_phi",CPUI_COPY);
    PcodeOp *indirect = fixture.make("indirect",CPUI_INDIRECT,2);
    PcodeOp *beforeTarget = fixture.make("before_target",CPUI_COPY,1);
    PcodeOp *afterIndirect = fixture.make("after_indirect",CPUI_COPY);
    PcodeOp *endNormal = fixture.make("end_normal",CPUI_COPY);

    fixture.tracked = beforeTarget;
    fixture.trackedInput = fd->newConstant(8,0x1234);
    fd->opSetInput(beforeTarget,fixture.trackedInput,0);
    fixture.trackedOutput = fd->newUniqueOut(8,beforeTarget);
    fd->opSetInput(indirect,fd->newConstant(8,0),0);
    fd->opSetInput(indirect,fd->newVarnodeIop(target),1);
    fixture.dump("created",false);

    fd->opInsertEnd(terminal,fixture.block);
    fd->opInsertEnd(target,fixture.block);
    fd->opInsertBegin(head,fixture.block);
    fd->opInsertBegin(phiA,fixture.block);
    fd->opInsertBegin(phiB,fixture.block);
    fd->opInsertAfter(afterPhi,phiB);
    fd->opInsertBefore(indirect,target);
    fd->opInsertBefore(beforeTarget,target);
    fd->opInsertAfter(afterIndirect,indirect);
    fd->opInsertEnd(endNormal,fixture.block);
    fixture.dump("inserted",true);

    fd->opUninsert(beforeTarget);
    fixture.dump("uninserted",true);

    fd->opInsertAfter(beforeTarget,head);
    fixture.dump("reinserted",true);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: op_insert_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/op_insert_1204.rs
use std::collections::HashMap;
use std::sync::Arc;

use rugra::address::Address;
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;

struct Fixture {
    fd: Funcdata,
    block: Arc<std::sync::RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
    ops: Vec<PcodeOpRef>,
    names: HashMap<usize, &'static str>,
    tracked: Option<PcodeOpRef>,
    tracked_input: Option<Arc<std::sync::RwLock<rugra::varnode::Varnode>>>,
    tracked_output: Option<Arc<std::sync::RwLock<rugra::varnode::Varnode>>>,
}

impl Fixture {
    fn new() -> Self {
        let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 74);
        let block = fd.create_new_block();
        Self {
            fd,
            block,
            ops: Vec::new(),
            names: HashMap::new(),
            tracked: None,
            tracked_input: None,
            tracked_output: None,
        }
    }

    fn make(&mut self, name: &'static str, opcode: OpCode, inputs: usize) -> PcodeOpRef {
        let op = self.fd.new_op(inputs, Address::new(0x1000));
        self.fd.op_set_opcode(&op, opcode);
        self.names.insert(Arc::as_ptr(&op.0) as usize, name);
        self.ops.push(op.clone());
        op
    }

    fn order(&self, ops: &[PcodeOpRef]) -> String {
        ops.iter()
            .filter_map(|op| self.names.get(&(Arc::as_ptr(&op.0) as usize)).copied())
            .collect::<Vec<_>>()
            .join(",")
    }

    fn parents(&self) -> String {
        self.ops
            .iter()
            .filter_map(|op| {
                self.names.get(&(Arc::as_ptr(&op.0) as usize)).map(|name| {
                    let parent =
                        op.0.read()
                            .unwrap()
                            .parent
                            .as_ref()
                            .and_then(std::sync::Weak::upgrade)
                            .map(|parent| parent.read().unwrap().get_index())
                            .unwrap_or(-1);
                    format!("{name}:{parent}")
                })
            })
            .collect::<Vec<_>>()
            .join(",")
    }

    fn op_state(&self) -> String {
        use rugra::op::pcodeop_flags;

        self.ops
            .iter()
            .filter_map(|op| {
                self.names.get(&(Arc::as_ptr(&op.0) as usize)).map(|name| {
                    let op = op.0.read().unwrap();
                    format!(
                        "{name}:{}/{}/{}/{}/{}/{}/{}/{}",
                        op.get_eval_type(),
                        u8::from(op.is_branch()),
                        u8::from(op.is_call()),
                        u8::from(op.is_marker()),
                        u8::from(op.is_bool_output()),
                        u8::from(op.is_flow_break()),
                        u8::from((op.flags & pcodeop_flags::CODEREF) != 0),
                        u8::from(
                            (op.flags
                                & (pcodeop_flags::MARKER
                                    | pcodeop_flags::NONPRINTING
                                    | pcodeop_flags::NORETURN))
                                != 0
                        ),
                    )
                })
            })
            .collect::<Vec<_>>()
            .join(",")
    }

    fn dump(&self, stage: &str, include_order: bool) {
        let tracked = self.tracked.as_ref().unwrap();
        let tracked_guard = tracked.0.read().unwrap();
        let parent = tracked_guard
            .parent
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|parent| parent.read().unwrap().get_index())
            .unwrap_or(-1);
        let tracked_dead = tracked_guard.is_dead();
        let tracked_order = tracked_guard.start.get_order();
        drop(tracked_guard);

        let block_ops = self.block.read().unwrap().get_ops();
        let block_flags = self.block.read().unwrap().get_flags();
        let input_descend = self
            .tracked_input
            .as_ref()
            .unwrap()
            .read()
            .unwrap()
            .descend
            .iter()
            .filter(|weak| weak.upgrade().is_some())
            .count();
        let output_def = self
            .tracked_output
            .as_ref()
            .unwrap()
            .read()
            .unwrap()
            .def
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|definition| Arc::ptr_eq(&definition, &tracked.0))
            .unwrap_or(false);

        print!(
            "{stage}|block={}|alive={}|dead={}|parents={}|opstate={}|flags={block_flags}|tracked_dead={}|tracked_parent={parent}|tracked_input_desc={input_descend}|tracked_output_def={}",
            self.order(&block_ops),
            self.order(&self.fd.obank.alivelist),
            self.order(&self.fd.obank.deadlist),
            self.parents(),
            self.op_state(),
            u8::from(tracked_dead),
            u8::from(output_def),
        );
        if include_order {
            print!("|tracked_order={tracked_order}");
        }
        println!();
    }
}

fn main() {
    let mut fixture = Fixture::new();
    let terminal = fixture.make("terminal", OpCode::CPUI_BRANCHIND, 0);
    let target = fixture.make("target", OpCode::CPUI_STORE, 0);
    let head = fixture.make("head", OpCode::CPUI_COPY, 0);
    let phi_a = fixture.make("phi_a", OpCode::CPUI_MULTIEQUAL, 0);
    let phi_b = fixture.make("phi_b", OpCode::CPUI_MULTIEQUAL, 0);
    let after_phi = fixture.make("after_phi", OpCode::CPUI_COPY, 0);
    let indirect = fixture.make("indirect", OpCode::CPUI_INDIRECT, 2);
    let before_target = fixture.make("before_target", OpCode::CPUI_COPY, 1);
    let after_indirect = fixture.make("after_indirect", OpCode::CPUI_COPY, 0);
    let end_normal = fixture.make("end_normal", OpCode::CPUI_COPY, 0);

    fixture.tracked = Some(before_target.clone());
    let tracked_input = fixture.fd.new_constant(8, 0x1234);
    fixture
        .fd
        .op_set_input(&before_target, tracked_input.clone(), 0);
    let tracked_output = fixture.fd.new_unique_out(8, &before_target);
    fixture.tracked_input = Some(tracked_input);
    fixture.tracked_output = Some(tracked_output);
    let indirect_zero = fixture.fd.new_constant(8, 0);
    fixture.fd.op_set_input(&indirect, indirect_zero, 0);
    let target_iop = fixture.fd.new_varnode_iop(&target);
    fixture.fd.op_set_input(&indirect, target_iop, 1);
    fixture.dump("created", false);

    fixture.fd.op_insert_end(&terminal, &fixture.block);
    fixture.fd.op_insert_end(&target, &fixture.block);
    fixture.fd.op_insert_begin(&head, &fixture.block);
    fixture.fd.op_insert_begin(&phi_a, &fixture.block);
    fixture.fd.op_insert_begin(&phi_b, &fixture.block);
    fixture.fd.op_insert_after(&after_phi, &phi_b);
    fixture.fd.op_insert_before(&indirect, &target);
    fixture.fd.op_insert_before(&before_target, &target);
    fixture.fd.op_insert_after(&after_indirect, &indirect);
    fixture.fd.op_insert_end(&end_normal, &fixture.block);
    fixture.dump("inserted", true);

    fixture.fd.op_uninsert(&before_target);
    fixture.dump("uninserted", true);

    fixture.fd.op_insert_after(&before_target, &head);
    fixture.dump("reinserted", true);
}
